### PR TITLE
DROOLS-2160: Introduce tabs for KieSessionSelector and AuditLog

### PR DESCRIPTION
**Pull request by jomarko**
DROOLS-2160: Move config and audit to separate tabs

### CODE REVIEW

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by jomarko** [comment]
@manstis @Rikkola would you mind to have a look? As I said in the JIRA, this pr introduces separate tabs for scenario designer. One tab for configuration (kie base/session, allowed/prevented rules, date) and one tab for scenario result (audit log and fired rules).
@karreiro would you mind have a look on my usage of `elemental2`. I know you are kind of advanced user of it.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/ConfigWidget.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/ConfigWidget.java]: 2018

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePage.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePage.java]: 2018

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePageView.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePageView.java]: 2018

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePageViewImpl.html [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<!--
+  ~ Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePageViewImpl.html]: 2018

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/ScenarioKSessionSelector.java [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016 Red Hat, Inc. and/or its affiliates.
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/ScenarioKSessionSelector.java]: Now here I'm not sure if this should be 2016 or 2018. We can settle for 2017

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/ScenarioWidgetComponentCreator.java [hunk 6]
```diff
@@ -61,7 +60,7 @@ public ScenarioWidgetComponentCreator(final Caller<RuleNamesService> ruleNamesSe
         this.ruleNamesService = ruleNamesService;
     }
 
-    protected void reset(final ScenarioParentWidget scenarioWidget,
+    public void reset(final ScenarioParentWidget scenarioWidget,
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/ScenarioWidgetComponentCreator.java]: Format issue ^

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/test/java/org/drools/workbench/screens/testscenario/client/page/configuration/ExecutionWidgetTest.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+        executionWidget.show(executionTrace);
+
+        verify(dateConfigurationChoices).setSelectedIndex(0);
+        verify(dateTimePicker).setValue(null);
+    }
+
+    @Test()
+    public void testShowTimePickerUsedBefore() throws Exception {
+        final Date date= new Date();
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/test/java/org/drools/workbench/screens/testscenario/client/page/configuration/ExecutionWidgetTest.java]: Format issue ^

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/test/java/org/drools/workbench/screens/testscenario/client/utils/ScenarioUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import org.drools.workbench.models.testscenarios.shared.Scenario;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ScenarioUtilsTest {
+
+    @Test
+    public void testFindExecutionTrace() throws Exception {
+        final Scenario scenario = new Scenario();
+        final ExecutionTrace executionTrace =  new ExecutionTrace();
```
Comment (karreiro) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/test/java/org/drools/workbench/screens/testscenario/client/utils/ScenarioUtilsTest.java]: Code format ^

**Comment by jomarko** [comment]
@karreiro Hi, I refactored the `AuditLogTable` to use html templates, however I have troubles to style it, anything I add into `AuditLogTable.less` (not present in PR for now) file doesn't take effect, could you please give me some tip?

**Comment by karreiro** [comment]
@jomarko AFAIK, the only thing that we need to do when we have a  `.less` file is make its name the same of the View (eg.: a `PopUpView.java` and `PopUpView.less`.

Here's an example where the `.less` is working:
https://github.com/karreiro/drools-wb/blob/b711104c19541b6794c97f784b246c7905de2368/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/editor/page/accordion/GuidedDecisionTableAccordionView.java
https://github.com/karreiro/drools-wb/blob/b711104c19541b6794c97f784b246c7905de2368/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/editor/page/accordion/GuidedDecisionTableAccordionView.less

Ohh! And remember, when you change or create `.less` file, you need to restart the webapp.

**Comment by jomarko** [comment]
Thank you @karreiro, would you mind to review this one more time please?

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                             final HTMLDivElement firedRulesDiv,
+                             final AuditTable firedRulesTable,
+                             final AuditTable auditLogTable) {
+        this.root = root;
+        this.auditLogDiv = auditLogDiv;
+        this.firedRulesDiv = firedRulesDiv;
+        this.firedRulesTable = firedRulesTable;
+        this.auditLogTable = auditLogTable;
+
+        this.firedRulesDiv.appendChild(firedRulesTable.getElement());
```
Comment (manstis) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java]: I've yet to write a Elemental2 screen so am guessing here; but can you not add the Rules Table to the Audit Page using the HTML? If your ```@DataField``` annotations were against the two ```AuditTable``` fields and you removed the "containing" ```HTMLDivElement```s? What you've done here is in essence what ```errai-ui``` does.

Comment (manstis) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java]: ```AuditTable``` implements ```IsElement``` so should be "injectable" into the HTML.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditTable.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Inject
+    public AuditTable(final HTMLUListElement itemsContainer,
+                      final ManagedInstance<AuditTableItem> items,
+                      final Elemental2DomUtil elemental2DomUtil) {
+        this.itemsContainer = itemsContainer;
+        this.items = items;
+        this.elemental2DomUtil = elemental2DomUtil;
+
+        this.title = items.get();
```
Comment (manstis) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditTable.java]: Can we not setup the Header in the HTML, using an instance ```@DataField``` for a ```AuditTableItem```?

You could then have the HTML contain one "outer" ```DIV``` containing two children ```DIV```; one for Header and one for the items. You would not need to then "add back" the header in the ```showItems(...)``` method.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePage.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import org.drools.workbench.models.testscenarios.shared.Scenario;
+import org.drools.workbench.screens.testscenario.client.ScenarioParentWidget;
+import org.drools.workbench.screens.testscenario.client.resources.i18n.TestScenarioConstants;
+import org.jboss.errai.common.client.ui.ElementWrapperWidget;
+import org.uberfire.backend.vfs.Path;
+import org.uberfire.client.mvp.UberElemental;
+import org.uberfire.client.views.pfly.multipage.PageImpl;
+
+@Dependent
+public class KiePage extends PageImpl {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePage.java]: Warning over naming this ```KiePage``` and having it in ```drools-wb```.

It looks to be specific to Test Scenarios and hence I'd recommend changing the name accordingly.

```KiePage``` implies, to me at least, that it's a generic re-usable page for all KIE related UI work.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    public void showFiredRules(final ExecutionTrace executionTrace) {
+        firedRulesTable.showItems(Arrays.asList(executionTrace.getRulesFired()));
+    }
+
+    @Override
+    public void showAuditLog(final Set<String> auditLogMessages) {
+        auditLogTable.showItems(auditLogMessages);
```
Comment (manstis) [inline drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java]: IDK if the use of ```Set``` has changed at all in this refactoring; however I'd be inclined to convert to a ```List``` and sort it before passing to the View for display. WDYT?

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java (hunk 9 above)
Comment (manstis) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java]: @mbarkley Is my understanding correct? 

We should be able to have ```errai-ui``` _inject_ **any** ```IsElement``` into the HTML template?

Comment (mbarkley) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java]: @manstis Yes, you should be able to put the `AuditTable` as a `@DataField` and have the corresponding `data-field` element be a child of the `data-field="fired-rules"` element in your HTML template. This should work if the `AuditTable` is any kind of UI component (Errai UI `@Templated` bean, GWT Widget, Elemental2 IsElement, Errai DOM Wrapper IsElement).

In general the only reason you *might* ever need to manually construct parts of the DOM with Errai UI is for dynamic UI components (i.e. entries in a list or table, populated from somewhere else).

Comment (jomarko) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java]: Great, will apply.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditTable.java (hunk 10 above)
Comment (jomarko) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditTable.java]: Sounds good. I will do.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java (hunk 12 above)
Comment (jomarko) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java]: I think this is not good way, the data looks like [1]. I think if we want to improve audit log, we should return proper list directly form backend.

[1]
screenshot from 2018-01-24 22-58-20

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePage.java (hunk 11 above)
Comment (jomarko) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/configuration/KiePage.java]: Sounds reasonable, I will do.

File: drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java (hunk 12 above)
Comment (manstis) [reply drools-wb-screens/drools-wb-test-scenario-editor/drools-wb-test-scenario-editor-client/src/main/java/org/drools/workbench/screens/testscenario/client/page/audit/AuditPageViewImpl.java]: Ok. Fair enough.

**Comment by jomarko** [comment]
@manstis I added commit according to your review. Thanks for looking.
@karreiro sorry for disturbing, I would like to ask you for review too, I am going to cherry pick this into 7.5.x as this is part of Getting Started Experience requirement, targeted for RHDM-7.0.GA.

**Comment by manstis** [comment]
@jomarko KiePageTest probably need renaming otherwise looks good to me.

**Comment by jomarko** [comment]
@manstis well spotted, done.

**Comment by manstis** [comment]
@jomarko Please let me know when green and I'll merge (we don't need @mbarkley to approve)

**Pull request merged**